### ISSUE TITLE
AssemblyScript Public Meeting #41 - July 14th, 2021

### ISSUE THREAD

**Issue by torch2424** [AssemblyScript/core#68]
# AssemblyScript Public Meetings

Anyone who is contributing to the AssemblyScript project, building something with AssemblyScript, interested in the Assembly project or WebAssembly in general, is welcome to join! 😄

This meeting is focused on both fostering feedback from the AssemblyScript communitty and ecosystem, as well as for maintainers to discuss technical problems or achievements on AssemblyScript projects. :rocket: 

# Date and Time

AssemblyScript Public meetings occur every two weeks.

This public meeting will take place: **July 14th, 2021, 17:00 UTC (10:00 AM US PT, UTC -8)**

# General Agenda

* Agenda Items from comments left on this Github issue
  * **Please feel free to comment below to add anything to the agenda.**
* Additional in-meeting comments / discussion

Feel free to comment on this issue if you have any agenda items you would like to bring up. 
Meeting Notes will be placed on this issue for those who cannot make the meeting.

# Meeting Information

We will use the AssemblyScript discord for meetings: https://discord.gg/vSBbJ5T . We will meet in the "meeting" voice channel :+1: 

Notes will be taken by the host(s) ( @torch2424 and/or @willemneal ), and posted after the meeting. Notes are free to be edited through comments on the meeting notes at a later time.

**Comment by torch2424** [comment]
## Meeting Notes

* @torch2424 Things I’ve been doing
   * Github README Interview should be going up soon.
      * Highlighted the project, and one of the highlighted quotes is: 
         * “AssemblyScript gave me the freedom to pick up WebAssembly without having to learn an entirely new language.”
   * Working with @willemneal  on assemblyscript-json
   * Been writing a lot of AS documentation on the AS side of things (for Fastly Stuff).
      * Highlighted libraries like as-regex and assembly script-json
   * General as-bind maintenance, and need to open PR on as-wasi.
* @JairusSW JSON in Assembly script
   * how are dynamic arrays handled in assembly script-json
      * assembly script-json does: https://github.com/near/assemblyscript-json/blob/main/assembly/JSON.ts#L281 
      * We have a generic JS value that gets pushed onto the array
* @jtenner : Working on as-lunatic
   * https://github.com/lunatic-solutions/as-lunatic/pull/14 
   * Adding MultiValue support, need a code review at the link above on the AS side of things
      * @willemneal  said they would help with reviews
* @willemneal : Quick General things they have been working on:
   * Working on a few PRs to help with developing transforms
   * PRs
      * https://github.com/AssemblyScript/assemblyscript/pull/1964 
      * https://github.com/AssemblyScript/assemblyscript/pull/1979 
      * https://github.com/AssemblyScript/assemblyscript/pull/1977 
   * @willemneal  has had some Assembly script dependency weirdness that can happen.
      * @dcodeIO mentioned it may be because of weird npm peerDependencies
      * @jtenner mentioned they have hit this same issue, due to having AS installed globally vs. locally
         * Easy to get around, but definitely is there
      * @dcodeIO : I think this comes down to a fundamental problem with handling package.json and we need a convention on how to do this in AS
      * (Essentially, this discussion comes down to publishing AS transform libraries, and how do transform authors keep up with the compiler, while making sure transform AS versions don’t conflict)
         * Action Items:
            * @dcodeIO : Lets try out what we can solve with peerDependencies
            * @jtenner & @willemneal : If that doesn’t work, let’s go ahead and expose the compiler version in the transforms
               * We can work on this together
      * (Discussion between @willemneal  and @dcodeIO about Warnings and .tsignore)
* @battlelinegames : WASI & Web Discussions (NOTE: This is a bit of a heated subject in general, so please don’t quote my notes here. Please reach out to the folks here directly for their opinions. This discussion was very friendly and nice to have haha! :) )
   * Summary: It seems like we are going to have to choose if AS is going to be a Web First language, vs. a WASI first language
   * Question: If we had to choose as a group, which would we choose to be the the preference
      * @willemneal : I think we should be Wasm first
      * @torch2424 : I think if we had to choose browser APIs vs. WASI ones. I think I’d vote for browser use cases. BUT I say we support both as best as we possibly can
      * @battlelinegames : I think we should prioritize the web in times of conflict
   * @dcodeIO : Well, let’s say we depend entirely on JS, using JS strings and arrays. And then people using AS outside of the browser, they wouldn’t have this, and it wouldn’t work anymore for non-browser-like use cases. And the Stdlib would eventually go away
      * @willemneal  & @jtenner : I think that’s a step backwards. And is discouraging, we should definitely keep up with the stdlib, and support non-browser use cases
   * @dcodeIO : Of course I want to support both. And why we built a GC and a stdlib. We want to integrate with Non-web as well. 
   * @battlelinegames : So I guess I’m trying to understand where these specs are conflicting and we have to draw a line in the sand
      * @battlelinegames : We don’t think Interface types is compatible, with the GC, Type Reflection, etc.. ?
      * @dcodeIO : Well, we can essentially use externref to get a lot of things for free, without having to copy, so we wouldn’t really need to use Interface Types
      * @dcodeIO : Essentially the two paths are:
         * WASI: We double down on the stdlib, and we rewrite how a lot things are handled in AS internally (Particularly, replacing the string implementation) to work better for interface and other languages that support WASI well
         * Browser: We throw away the stdlib, and we import everything from a JS host using reference types and things
      * (Discussions about general Spec discussions, and how UTF-16 has a soft commitment. @dcodeIO emphasized that UTF-16 is not currently definitively supported)
      * (General discussions about how the spec proposals can play out, and how they would affect the language)
   * @torch2424 : Two thoughts
      * 1. I agree with @battlelinegames , I think we are looking a little too far ahead here, and we probably want to wait before making any big decisions, because a lot of these proposals still have a lot of work and phases ahead of them
      * 2. An idea for the worse case scenario, where we got a JS Host first solution. I would say we slim down what the compiler is responsible for, and eject out the stdlib into something that can be maintained by our sponsors, and the general WASI community, since WASI will need all of the JS re-implementations (Again, In the unlikely, worse case spec scenario for compiler authors)
   * @MaxGraey : I don’t think it’d be a good idea to change all of the internal representation to a reference based implementation for stdlib and things.
      * I also think it’s a bit early to make this decision. We’d probably want to wait until things land, and then test how much impact it has before starting all the work and making the huge shift.
   * (Discussion continued but I had to jump to another meeting)

**Comment by torch2424** [comment]
Closing in favor of: https://github.com/AssemblyScript/core/issues/69

Feel free to continue discussions from this meeting here!
